### ISSUE TITLE
Query on update the ADFS FederationMetadata.xml on the keycloak instead of delete and recreating the IDP config #24310

### ISSUE THREAD

**Issue by DREGALLA** [keycloak/keycloak#24424]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

saml

### Describe the bug

when we use identity provider( example: ADFS, Ping Identity, Okta,Azure etc ) for SSO, If the IDP certificate has expired and has been replaced , Does the FederationMetadata.xml on the keycloak for the IDP also need to be replaced ?

If yes how can we update the ADFS FederationMetadata.xml on the keycloak?

Keycloak version: 22.0.4

### Version

22.0.4

### Expected behavior

no need to replace FederationMetadata.xml on the keycloak for the IDP

### Actual behavior

Delete the IDP client , recreate  IDP client and  replace FederationMetadata.xml on the keycloak for the IDP

### How to Reproduce?

when we use identity provider( example: ADFS, Ping Identity, Okta,Azure etc ) for SSO, If the IDP certificate has expired and has been replaced, we need to delete the IDP client, recreate IDP client and replace FederationMetadata.xml on the keycloak for the IDP

This is not a best practice on production system.

**Comment by rmartinc** [comment]
@DREGALLA You are right, configuring a SAML provider with the `SAML entity descriptor` option just downloads the configuration at UI level, creation time. It performs the initial configuration for the provider and nothing else. The URL is not used anymore (it's not even saved in the provider configuration) and you have to update the configuration manually. We can improve that but this is how it works now.

So I'm changing this issue to enhancement. This is not a bug. Probably the certificate should be requested from time to time, when verification fails,... If anyone wants to work on this would be much appreciated. Moving to backlog.

**Comment by DREGALLA** [comment]
Thank you for sharing the information

**Comment by DREGALLA** [comment]
@rmartinc , Does the SAML protocol in Keycloak offer automatic certificate rollover support for ADFS?   IS IDP certificate gets renewed or auto-renewed before it expires in Keycloack automatically. Could you please give us brief details over it.

Thank you for helping us.

**Comment by rmartinc** [comment]
@DREGALLA No, as I I commented in the previous comment, for the moment there is no automatic rollover of the certificate for SAML identity providers. The metadata URL is just used when creation to auto-configure the provider but nothing else. The certificate should be changed manually, you have to get the new certificate in the IDP and set manually to the option `Validating X509 certificates`. You can use more that one cert in that option separated by commas. So you can configure in advance two certs (the old and the new one at the same time) to avoid failures at change time (and then delete the old one).

**Comment by BagalDipti** [comment]
https://faq.miniorange.com/knowledgebase/how-to-renew-upgrade-sp-certificate-in-the-saml-sso-plugin-with-keycloak-as-idp/ 

The above link suggests to replace the cert only when ‘Encrypt Assertion’ and ‘Client Signature Required’ is set ‘ON’

we don’t have these specific settings above under the SAML config menu in keycloak 6.0.  Also the link mentioned where we can import an encryption key and signing key.

But, on Key-cloak version 6.0.0 packaged with AADS,  the ‘Keys’ tab  is not available under the Client >> AADS Menu
It seems that we are not enabling SAML assertion encryption by default on Key-cloak.

Is this the reason why SAML SSO continues working although the cert has expired?

**Comment by omasseau** [comment]
Hello,

Just to exactly understand what this new 'Allow automatic download of SAML certifcates in the identity provider' functionnality does : 

Does is just allow to automatically download the certificates at the moment the import of the SAML metadata url is done (which is done manually when creating the Identity Provider in Keycloak) ?
OR
Does it automatically detect the changes of the certificates defined in the metadata when they change, in order to import them automatically (in case of the IDP certificates rollover) ?

To summarize: Do I still have to manually re-import the metadata url when there is a SAML certificates rollover ?

**Comment by rmartinc** [comment]
Hi @omasseau!

I think it is explained in the [release notes](https://www.keycloak.org/docs/latest/release_notes/index.html#automatic-certificate-management-for-saml-identity-providers). Please client the link to the descrption of the new options too. I think the answer is you can do both depending what you configure. 

If you setup the `Metadata descriptor URL` with the URL and the `Use metadata descriptor URL`to ON, the certificates are automatically downloaded from the URL and cached when needed (if direct certificate is requested it's downloaded when not found, if all certificates are requested, they are refreshed by max time).

If you just setup `Metadata descriptor URL` with the URL but  `Use metadata descriptor URL` to OFF, there is no automatic download and you need to re-import the certificates manually. You can do that with the admin console in the action combo now.
